### PR TITLE
scrape: detect duplicate series after metric relabeling

### LINKED ISSUE #12255: Duplicates leak to remote-write
### What did you do?

Per docs prometheus works in this order:
- scrapes target 
- do relabel
- drop duplicates (logs them: https://github.com/prometheus/prometheus/issues/11725)
- ingest the rest, there cannot be duplicates in tsdb by design
- replicate local tsdb from WAL via remote-write to target endpoint

Unfortunately we see that duplicates also leak to remote-write.
Way to reproduce:
- We will start prometheus, which would scrape nginx and remote-write to mimir:
```yaml
version: '2'
services:

  prometheus:
    container_name: prometheus
    hostname: prometheus
    image: prom/prometheus:v2.43.0
    command:
      - --config.file=/cfg/prometheus.yml
    volumes:
      - ./prometheus.yml:/cfg/prometheus.yml
    ports:
      - '9090:9090'

  nginx:
    container_name: nginx
    hostname: nginx
    image: nginx
    volumes:
      - ./metrics:/usr/share/nginx/html/metrics
      - ./metrics2:/usr/share/nginx/html/metrics2
    ports:
      - '8081:80'
      
  mimir:
    container_name: mimir
    hostname: mimir
    image: grafana/mimir:2.7.1
    command:
      - -auth.multitenancy-enabled=false
      - -ingester.ring.replication-factor=1
    ports:
      - '8080:8080'
```
- prometheus config:
```yaml
global:
  scrape_interval: 1s
scrape_configs:
  - job_name: own
    static_configs:
      - targets:
          - prometheus:9090
          - mimir:8080

  - job_name: duplicates
    static_configs:
      - targets:
          - nginx:80

remote_write:
  - url: http://mimir:8080/api/v1/push
```
- duplicate metrics:
```sh
$ for i in `seq 100000`; do echo "same_metric_name $i"; done > metrics
$ head metrics
same_metric_name 1
same_metric_name 2
same_metric_name 3
same_metric_name 4
same_metric_name 5
same_metric_name 6
same_metric_name 7
same_metric_name 8
same_metric_name 9
same_metric_name 10
```
- start this via docker-compose, and let's see prometheus metrics:
```ini
same_metric_name{instance="nginx:80", job="duplicates"} 1 # all lines deduplicated to just one metric, correct
prometheus_tsdb_head_series{instance="prometheus:9090", job="own"} 5169 # no 100k series in tsdb, correct
scrape_samples_scraped{instance="nginx:80", job="duplicates"} 100000 # hm, ok
scrape_samples_post_metric_relabeling{instance="nginx:80", job="duplicates"} 100000 # shouldn't it be just 1 here?
# sending speed via remote-write in samples/s
rate(prometheus_remote_storage_samples_total[1m]) 102902 # why? there are only 5k samples in head
# mimir failed samples/s
rate(cortex_ingester_ingested_samples_failures_total[1m]) 6883.5292321782945 # that actually matters, see below
```
- logs on mimir side:
```
mimir       | ts=2023-04-12T18:02:44.403379149Z caller=push.go:130 level=error user=anonymous msg="push error" err="rpc error: code = Code(400) desc = failed pushing to ingester: user=anonymous: the sample has been rejected because another sample with the same timestamp, but a different value, has already been ingested (err-mimir-sample-duplicate-timestamp). The affected sample has timestamp 2023-04-12T18:02:44.39Z and is from series {__name__=\"same_metric_name\", instance=\"nginx:80\", job=\"duplicates\"}"
```
That seems as a bug on prometheus side for me. If there cannot be duplicates in tsdb, how do they appear when tsdb is being replicated via remote-write?

### What did you expect to see?

I would expect that all duplicates issues would be handled on prometheus side, and then only data passed all validations would be replicated further via remote-write.

### What did you see instead? Under which circumstances?

That does not sound like a big issue, but unfortunately it is harmful for people writing to mimir in distributor/ingester mode (i believe thanos-router/receive scheme would have the same issue).
- one of tenants got around 20 targets in his prometheus, exposing 400k duplicated metrics each. Not a big deal as we can ingest 9M ok-series easily
- any non-ok remote-write request starts to increase latency on incoming processing in mimir. In this case it has been 9M tainted samples, so large amount of incoming packets start to get 4xx
- usual incoming latency in our cluster 10-30ms, in this case it spiked to 1-3s
- mimir `distributor` component reads incoming http request to memory, parse remote-write proto and then replicates request to `ingesters`, awaits quorum number write confirmation and only then answers to incoming http request
- in case of increased latency it has to hold 3s of incoming 150Mbit/s flow in memory, it does back-pressure with currently established connections from prometheuses, but proms just spawn new shards if current slows down. Customers manage their prom configs, we cannot dictate `remote_write.queue_config` to them
- `distributors` start massively die of OOM, affects other. It even has context_deadline_timeout of 2s, so if we scale mem and replicas quickly, we anyway getting massive amounts of 5xx because ingester latency grows even higher than that

### System information

_No response_

### Prometheus version

_No response_

### Prometheus configuration file

_No response_

### Alertmanager version

_No response_

### Alertmanager configuration file

_No response_

### Logs

_No response_

### PR DESCRIPTION
**Pull request by Rajkaran-122**
Fixes #12255

## Summary

This change moves duplicate series detection to occur after metric relabeling.

Previously, duplicate detection in the scrape loop was based on the parsed metric before metric relabeling. If metric relabeling transformed multiple scraped metrics into the same final label set, duplicate detection did not occur until the samples reached the TSDB appender.

This change performs duplicate detection using the post-relabel label set by tracking label-set hashes during each scrape iteration, allowing duplicates introduced by metric relabeling to be detected before appending samples to TSDB.

## Changes

- Track post-relabel label-set hashes during each scrape iteration.
- Perform duplicate detection after metric relabeling in both V1 and V2 scrape appenders.
- Clear the per-scrape hash tracker at the end of each scrape iteration.
- Add a regression test covering duplicate label sets introduced by metric relabeling.

#### Release notes for end users

```release-notes
[BUGFIX] Scrape: Detect duplicate series after metric relabeling before appending samples.
```

### CODE REVIEW

**Comment by Rajkaran-122** [comment]
The remaining CI failure is in TestHeadCompactionWhileScraping under cmd/prometheus, which is outside the scope of this PR. The logs show TSDB compactions occurring, but the test times out waiting for its expected condition. The scrape-related tests for this change completed successfully. I'm happy to investigate further if you think it's related to my changes.

**Comment by bboreham** [comment]
Thanks for this; could you run the relevant benchmarks and post the comparison using `benchstat` please.

Did you consider hash collisions?  (It's not essential, but would at least need to be thought about and noted)

**Comment by Rajkaran-122** [comment]
> Thanks for this; could you run the relevant benchmarks and post the comparison using `benchstat` please.
> 
> Did you consider hash collisions? (It's not essential, but would at least need to be thought about and noted)

Thanks for the feedback   @bboreham sir ! I investigated the Windows CI failure and it appears unrelated to this change. TestManagerSTZeroIngestion uses an explicit TimestampMs with honor_timestamps: true, while the duplicate detection introduced by this PR is only exercised when the parsed timestamp is absent. The 1 ms mismatch comes from the existing OpenMetrics float-seconds to milliseconds conversion, so I don't think this PR should modify or weaken that assertion.

I also reviewed the duplicate tracker and agree that treating a 64-bit hash as identity is not ideal. I'll update the implementation to handle hash collisions by verifying matching hashes with labels.Equal, following the TSDB approach, and add a regression test using the existing collision fixture.

I'll also run the relevant BenchmarkScrapeLoopAppend benchmarks (V1/V2, with and without storage) and post the benchstat comparison once I have stable results.

**Comment by Rajkaran-122** [comment]
Thanks for the feedback, @bboreham sir!

I've updated the implementation to handle hash collisions using the TSDB approach (labels.Equal with conflict buckets), added a regression test for the existing collision fixture, and ran the requested benchmarks. benchstat shows an overall geomean difference of about +1.74%, with no significant performance regression. All CI checks are now passing. Happy to make any further adjustments if needed.

**Comment by bboreham** [comment]
> post the comparison using benchstat please

Post the actual output.

**Comment by Rajkaran-122** [comment]
> Post the actual output.

Thanks @bboreham sir ! Here's the benchstat comparison:

var buf strings.Builder 
buf.WriteString(ls1.Get(model.MetricNameLabel))
buf.WriteString("{")
- first = true
+ first := true
ls1.Range(func(l labels.Label) {


The results show no meaningful performance regression for this change.

**Comment by Rajkaran-122** [comment]
@pedro-stanaka Thanks @bboreham sir for taking the time to review this. I've addressed the requested feedback and updated the PR accordingly. If you notice anything else that should be improved, I'm happy to make further changes.

**Comment by Rajkaran-122** [comment]
@bboreham , @pedro-stanaka  Hi! Just a gentle follow-up. The requested feedback has been addressed, the benchmark output has been posted, hash collision handling has been updated, and all checks are now passing. I'd appreciate a review whenever you have time. Thank you!

**Comment by Rajkaran-122** [comment]
@bboreham sir,
Benchmark results (6 runs, -benchtime 1s, -cpu 2):

Compared with main, the updated implementation shows a substantial overall improvement:

- sec/op: -48.59% geomean
- B/op: -26.54% geomean
- allocs/op: -32.26% geomean

The largest improvements are seen in the larger 237FamsAllTypes
workloads, with several formats showing ~67–74% lower execution time.

There is a regression for the small 1Fam2000Gauges/PromText workload
(+49.36% sec/op, +38.33% B/op, +24.62% allocs/op), which should be
considered separately from the larger-workload improvements.

Benchmarked on Windows/amd64, 11th Gen Intel Core i7-1185G7 @ 3.00GHz,
with 6 samples per benchmark.

> > post the comparison using benchstat please
> 
> Post the actual output.

output
goos: windows
goarch: amd64
pkg: github.com/prometheus/prometheus/scrape
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz

                                                                                                              │ main_v1_storage.txt │           pr_v1_storage.txt           │
                                                                                                              │       sec/op        │     sec/op      vs base               │
ScrapeLoopAppend/withStorage=true/appV2=false/appendMetadataToWAL=false/data=1Fam2000Gauges/fmt=PromText-2           971.7µ ± 110%   1451.4µ ± 142%  +49.36% (p=0.026 n=6)
ScrapeLoopAppend/withStorage=true/appV2=false/appendMetadataToWAL=false/data=1Fam2000Gauges/fmt=OMText-2             3.381m ±  69%    1.403m ± 130%        ~ (p=0.180 n=6)
ScrapeLoopAppend/withStorage=true/appV2=false/appendMetadataToWAL=false/data=1Fam2000Gauges/fmt=PromProto-2          1.482m ± 196%    1.388m ±   9%        ~ (p=0.132 n=6)
ScrapeLoopAppend/withStorage=true/appV2=false/appendMetadataToWAL=false/data=237FamsAllTypes/fmt=PromText-2          5.247m ±   2%    1.618m ±   9%  -69.16% (p=0.002 n=6)
ScrapeLoopAppend/withStorage=true/appV2=false/appendMetadataToWAL=false/data=237FamsAllTypes/fmt=OMText-2            4.837m ±   9%    1.592m ±  15%  -67.10% (p=0.002 n=6)
ScrapeLoopAppend/withStorage=true/appV2=false/appendMetadataToWAL=false/data=237FamsAllTypes/fmt=PromProto-2         5.236m ±  25%    1.388m ±  13%  -73.49% (p=0.002 n=6)
ScrapeLoopAppend/withStorage=true/appV2=false/appendMetadataToWAL=true/data=1Fam2000Gauges/fmt=PromText-2            3.615m ±  11%    1.080m ±  14%  -70.13% (p=0.002 n=6)
ScrapeLoopAppend/withStorage=true/appV2=false/appendMetadataToWAL=true/data=1Fam2000Gauges/fmt=OMText-2              3.717m ±  10%    1.055m ±   8%  -71.62% (p=0.002 n=6)
ScrapeLoopAppend/withStorage=true/appV2=false/appendMetadataToWAL=true/data=1Fam2000Gauges/fmt=PromProto-2           5.141m ±   7%    1.320m ±  15%  -74.33% (p=0.002 n=6)
ScrapeLoopAppend/withStorage=true/appV2=false/appendMetadataToWAL=true/data=237FamsAllTypes/fmt=PromText-2           1.794m ± 210%    1.754m ±   7%        ~ (p=0.818 n=6)
ScrapeLoopAppend/withStorage=true/appV2=false/appendMetadataToWAL=true/data=237FamsAllTypes/fmt=OMText-2             1.591m ±  10%    1.500m ±  20%        ~ (p=0.485 n=6)
ScrapeLoopAppend/withStorage=true/appV2=false/appendMetadataToWAL=true/data=237FamsAllTypes/fmt=PromProto-2          1.437m ±  19%    1.565m ±   7%        ~ (p=0.093 n=6)

geomean                                                                                                              2.745m           1.412m         -48.59%

**Comment by Rajkaran-122** [comment]
@bboreham sir please check the output.

**Comment by Rajkaran-122** [comment]
@sepich sir ,please review this pr

File: scrape/scrape.go
Comment (sepich) [inline scrape/scrape.go:1707]: `iterDone` is the only place `seriesHashes`/`seriesHashConflicts` get cleared, and the deferred call here returns early when `err != nil`. So any errored scrape: `sample_limit` exceeded, label limit exceeded, `errNameLabelMandatory`, invalid metric name leaves the whole previous scrape's hashes in the map. Next scrape flags every series as `ErrDuplicateSampleForTimestamp` and drops it.

File: scrape/scrape.go [hunk 1]
```diff
@@ -989,10 +997,38 @@ func (c *scrapeCache) iterDone(flushCache bool) {
 	// Swap current and previous series then clear the new current, to save allocations.
 	c.seriesPrev, c.seriesCur = c.seriesCur, c.seriesPrev
 	clear(c.seriesCur)
+	clear(c.seriesHashes)
+	clear(c.seriesHashConflicts)
```
Comment (sepich) [inline scrape/scrape.go]: `map[uint64]labels.Labels` holds one entry per series per scrape. A 1M-series target adds tens of MB, and `clear()` keeps bucket capacity forever. A single spike is permanent. Every other cache in `scrapeCache` is covered by the flush heuristic in `iterDone`; `len(seriesHashes)` isn't in the `count` computed there, so `targetScrapeCacheFlushForced` under-reports.

File: scrape/scrape.go [hunk 2]
```diff
@@ -1699,7 +1735,7 @@ loop:
 		if sl.cache.getDropped(met) {
 			continue
 		}
-		ce, seriesCached, seriesAlreadyScraped := sl.cache.get(met)
+		ce, seriesCached, _ := sl.cache.get(met)
```
Comment (sepich) [inline scrape/scrape.go:1764]: All `sl.cache.get()` calls now drop third value? Drop it from the signature then?

File: scrape/scrape.go [hunk 3]
```diff
@@ -989,10 +997,38 @@ func (c *scrapeCache) iterDone(flushCache bool) {
...
+	clear(c.seriesHashes)
+	clear(c.seriesHashConflicts)
 
 	c.iter++
 }
 
+// isSeriesHashDuplicate reports whether lset was already seen in the current
+// scrape iteration. It follows the same approach as TSDB's seriesHashmap:
+// a primary map holds the first label set per hash, and a lazily-allocated
+// conflict map handles the rare case of different label sets sharing a hash.
```
Comment (sepich) [inline scrape/scrape.go:1024]: The function mutates both maps, but docs says it only "reports whether".
Might worth updating the doc.

File: scrape/scrape_test.go [hunk 4]
```diff
@@ -6834,6 +6681,110 @@ func testScrapeLoopSeriesAddedDuplicates(t *testing.T, appV2 bool) {
...
+		if !first {
+			buf.WriteString(",")
+		}
+		buf.WriteString(l.Name)
+		buf.WriteString(`="`)
+		buf.WriteString(l.Value)
+		buf.WriteString(`"`)
+		first = false
+	})
+	buf.WriteString("} 2\n")
```
Comment (sepich) [inline scrape/scrape_test.go:6817]: Looks like copy-paste of `ls1`, might worth extracting to one func

File: scrape/scrape_test.go [hunk 5]
```diff
@@ -6834,6 +6681,110 @@ func testScrapeLoopSeriesAddedDuplicates(t *testing.T, appV2 bool) {
...
+	total, added, seriesAdded, err := app.append([]byte("test_metric{drop=\"a\"} 1\ntest_metric{drop=\"b\"} 2\n"), "text/plain", time.Time{})
+	require.NoError(t, err)
+	require.NoError(t, app.Commit())
+	require.Equal(t, 2, total)
+	require.Equal(t, 2, added)
+	require.Equal(t, 1, seriesAdded) // Only one series should be added.
+	require.Equal(t, 1.0, prom_testutil.ToFloat64(sl.metrics.targetScrapeSampleDuplicate))
+}
+
+func labelsWithHashCollision() (labels.Labels, labels.Labels) {
```
Comment (sepich) [inline scrape/scrape_test.go:6712]: This is a third copy 
https://github.com/prometheus/prometheus/blob/ca578101aff0022c64d8a47827de6fcca0a168f2/tsdb/agent/series_test.go#L78

**Review by sepich — APPROVED** [review]
> @sepich sir ,please review this pr

I'm not from the Team, so my review means nothing


> The largest improvements are seen in the larger 237FamsAllTypes
workloads, with several formats showing ~67–74% lower execution time.

Sorry, can't confirm that. The change only adds a map lookup + insert per sample, it can't be faster.
Your table has ±210% - which is noise due to busy windows laptop. Here is my repro run on base 40ea54d0b vs fcb82b6f0, `-count=6 -cpu 2 -benchtime 1s`:
```
                                          │   base     │           PR
data=1Fam2000Gauges/fmt=PromText  sec/op   392.3µ ± 1%   440.2µ ± 2%  +12.21% (p=0.002)
                                  B/op     2.830Ki ± 2%  3.060Ki ± 2%  +8.13% (p=0.002)
                                  allocs   22.00 ± 5%    23.00 ± 0%    +4.55% (p=0.015)

data=237FamsAllTypes/fmt=PromText sec/op   602.3µ ± 3%   649.9µ ± 6%   +7.90% (p=0.002)
                                  B/op     3.587Ki ± 5%  3.877Ki ± 4%  +8.10% (p=0.004)
```

**Comment by bboreham** [comment]
Looks like #19014 is in the same area?
